src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -172,29 +180,86 @@ private Continuable createContinuable(CpsThread currentThread, CpsCallableInvoca
     }
 
     @Override
-    public synchronized Collection<StepExecution> getCurrentExecutions() {
-        if (thread==null)   return Collections.emptyList();
-
-        StepExecution s = thread.getStep();
-        if (s!=null)        return Collections.singleton(s);
-        else                return Collections.emptyList();
+    public Collection<StepExecution> getCurrentExecutions() {
+        CpsThread t;
+        synchronized (this) {
+            t = thread;
+            if (t == null) {
+                return Collections.emptySet();
+            }
+        }
+        final SettableFuture<Collection<StepExecution>> result = SettableFuture.create();
+        t.getExecution().runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
+            @Override public void onSuccess(CpsThreadGroup g) {
+                try {
+                    List<StepExecution> executions = new ArrayList<>();
+                    // cf. trick in CpsFlowExecution.getCurrentExecutions(true)
+                    Map<FlowHead, CpsThread> m = new LinkedHashMap<>();
+                    for (CpsThread t : g.threads.values()) {
+                        m.put(t.head, t);
+                    }
+                    for (CpsThread t : m.values()) {
+                        // TODO seems cumbersome to have to go through the flow graph to find out whether a head is a descendant of ours, yet FlowHead does not seem to retain a parent field
```

Comment:
> adding an `enclosingBlockStartNodeId` field to `FlowNode`

Well it could be `transient` since it recomputable based on existing information. Anyway that would not help here I think, because we are starting with thread information, which does not completely align with flow graph information.

(For `getCurrentExecutions`—which no one besides my new test ever calls—it would probably suffice: you could filter `FlowExecution.currentHeads` to those enclosed by this body. But for `cancel` we need the actual `CpsThread`s, since these have special behavior for `stop` even when not inside a `Step`.)